### PR TITLE
Support installation of clangd server.

### PR DESCRIPTION
**Pull request by sam-mccall**
This supports downloading the latest zip release from github (on
supported platforms), extracting it within the extension's 'global storage',
and running it from there.

Several flows are supported:
  - install triggered via command palette
  - check for updates triggered via command palette
  - check for updates automatically on startup (OFF by default in this patch)
  - offer to install if we can't find clangd (ON by default in this patch)

### CODE REVIEW

**Comment by sam-mccall** [comment]
Demo of install flow: https://dump.video/i/RSJf9j.mp4
Demo of update flow + "installed but not selected" condition: https://i.imgur.com/wu0u5VN.mp4

I do want to work out how to make this testable, and may even reusable with coc.nvim. But wanted to get some feedback on the UI decisions and structure too.

File: src/install.ts [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+// We have an apparently-valid clangd (`clangdPath`), check for updates.
+export async function checkUpdates(clangdPath: string, requested: boolean,
+                                   context: vscode.ExtensionContext) {
+  // Gather all the version information to see if there's an upgrade.
+  try {
+    var release = await Github.latestRelease();
+    var asset = Github.chooseAsset(release);
+    var upgrade = await Version.upgrade(release, clangdPath);
+  } catch (e) {
+    console.log("Failed to check for clangd update: ", e);
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply src/install.ts:86]: In general yes, but console.log actually shows the object in the console (e.g. you can expand it and inspect its properties)

File: src/install.ts [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+export function chooseAsset(release: Github.Release): Github.Asset|null {
+  const variants: {[key: string]: string} = {
+    "win32" : "windows",
+    "linux" : "linux",
+    "darwin" : "mac",
+  };
+  const variant = variants[os.platform()];
+  // 32-bit vscode is still common on 64-bit windows, so don't reject that.
+  if (variant && (os.arch() == "x64" || variant == "windows"))
+    return release.assets.find(a => a.name.indexOf(variant) >= 0)
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply src/install.ts:154]: Added. But "throw" triggers automatic semicolon insertion, so this is "just" style

File: src/install.ts [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+export async function upgrade(release: Github.Release, clangdPath: string) {
+  const releasedVer = released(release);
+  const installedVer = await installed(clangdPath);
+  if (!rangeGreater(releasedVer, installedVer))
+    return null;
+  return {old : installedVer.raw, new : releasedVer.raw};
+}
+
+// Get the version of an installed clangd binary using `clangd --version`.
+async function installed(clangdPath: string): Promise<semver.Range> {
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply src/install.ts:309]: Yeah, the intent is for the caller to catch these errors. Adding a comment.

File: src/install.ts [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+// Functions to finish an install workflow, informing the user.
+namespace End {
+// The configuration is not valid, the user should install clangd manually.
+export async function showInstallationHelp(message: string) {
+  const url = "https://clangd.llvm.org/installation.html";
+  message += `\nSee ${url} for help.`
+  if (await vscode.window.showErrorMessage(message, "Open website"))
+  vscode.env.openExternal(vscode.Uri.parse(url));
```
[the remark this answers is not in the document]
Comment (sam-mccall) [reply src/install.ts:167]: this is clang-format's fault... it's because of the missing semicolon on `message += ...`, but I'm not sure why.

**Comment by sam-mccall** [comment]
(I'd still like to take a look at the structure to see how to make this reusable for coc and for testing)

File: src/install.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  context.subscriptions.push(vscode.commands.registerCommand(
+      'clangd.install', async () => common.installLatest(ui)));
+  context.subscriptions.push(vscode.commands.registerCommand(
+      'clangd.update', async () => common.checkUpdates(true, ui)));
+  const status = await common.prepare(ui, cfg.get<boolean>('checkUpdates'));
+  return status.clangdPath;
+}
+
+class UI {
+  constructor(private context: vscode.ExtensionContext,
```
Comment (hokein) [inline src/install.ts:22]: I was going to say the parameters are unused, then realized the TS compiler will generate the class members for us.

Comment (sam-mccall) [reply src/install.ts:22]: Yeah, it's weird syntax but pretty handy...

**Pull request merged**